File: exercises/04-retries-and-errors/solution/activities.py
import requests
from temporalio import activity
from temporalio.exceptions import ApplicationError

ADNET = "http://localhost:9999"


@activity.defn
def publish_post(channel: str) -> str:
    attempt = activity.info().attempt
    activity.logger.info("Publishing to %s (attempt %d)", channel, attempt)
    resp = requests.post(f"{ADNET}/publish/{channel}", json={"campaign": "summer-splash"}, timeout=5)
    if 400 <= resp.status_code < 500:
        # A 4xx is the ad network saying "no, and asking again won't help."
        # non_retryable=True tells Temporal not to waste retries on it.
        raise ApplicationError(
            f"{channel} rejected the post: {resp.json().get('error', resp.text)}",
            type="ChannelPolicyError",
            non_retryable=True,
        )
    resp.raise_for_status()  # 5xx raises here -> retryable by default
    return resp.json()["placement_id"]

File: tracks/typescript/exercises/04-retries-and-errors/solution/activities.ts
import { activityInfo } from '@temporalio/activity';
import { ApplicationFailure } from '@temporalio/common';

const ADNET = 'http://localhost:9999';

export async function publishPost(channel: string): Promise<string> {
  console.log(`Publishing to ${channel} (attempt ${activityInfo().attempt})`);
  const response = await fetch(`${ADNET}/publish/${channel}`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ campaign: 'summer-splash' }),
    signal: AbortSignal.timeout(5_000),
  });
  if (response.status >= 400 && response.status < 500) {
    const detail = (await response.json()) as { error?: string };
    throw ApplicationFailure.nonRetryable(
      `${channel} rejected the post: ${detail.error ?? response.statusText}`,
      'ChannelPolicyError',
    );
  }
  if (!response.ok) {
    throw new Error(`${response.status} ${response.statusText}: ${await response.text()}`);
  }
  const body = (await response.json()) as { placement_id: string };
  return body.placement_id;
}
